### PR TITLE
add mypy-based validation, drop syntactic check on return type

### LINKED ISSUE #535: Syntactic return type check is imprecise and should be disabled/replaced with a proper type checking pass
Inside `EncodableCallable` we have a validation process wherein we check that the synthesized function has the right type signature. This check is done syntactically on the ast of the annotation.

In practice this is causing LLM calls to fail unnecessarily:
```python
def foo() -> Optional[dict]:
   pass
```

Will raise an exception of the LLM generates a program dict | None however mypy would not complain. We should either drop this syntactic checking, or switch to a proper mypy based validation.

@matthew9671 can produce a reproducible example

### PR DESCRIPTION
**Pull request by kiranandcode**
This PR updates the internals of `EncodableCallable` such that `mypy` based typechecking is done on the source code generated by the LLM. The code uses the `ctx: Mapping[str, Any]` to inject appropriate imports and stubs such that the code should type check. 

This replaces and removes the syntactic check on the ast on the return type where previously we were just syntactically checking the annotations on the returned function which had no guarantee of correctness.

The core interface to the type checking is through this function in `effectful.handlers.llm.type_checking`:

```python

def typecheck_source(
    module: ast.AST,
    ctx: typing.Mapping[str, Any],
    expected_params: list[type] | None,
    expected_return: type,
) -> None:
    """Type-check synthesized module code against expected signature and context.
    Builds a full source with prelude (ctx bindings as type stubs), the module body,
    and a postlude that assigns the function to an expected Callable type so mypy
    validates the signature. Raises ValueError with mypy output on type errors.
    """
```

Closes #535

### CODE REVIEW

File: effectful/handlers/llm/type_checking.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+    Returns:
+        (annotation_source, import lines e.g. ["from typing import Any"], type alias lines e.g. ["type X = int"])
+    """
+    imports: list[str] = []
+    type_alias_lines: list[str] = []
+    ann = _collect(ty, imports, type_alias_lines)
+    return ann, list(dict.fromkeys(imports)), type_alias_lines
+
+
+def typecheck_source(
```
Comment (eb8680) [inline effectful/handlers/llm/type_checking.py]: Let's move this to `evaluation.py` to be with the other steps

File: effectful/handlers/llm/type_checking.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+            ann_str, imps, type_aliases = _type_to_annotation_str(
+                typing.cast(type, inferred)
+            )
+            all_imports.extend(imps)
+            all_type_aliases.extend(type_aliases)
+            prelude_lines.append(f"{name}: {ann_str}")
+    return prelude_lines, all_imports, all_type_aliases
+
+
+def _type_to_annotation_str(ty: type) -> tuple[str, list[str], list[str]]:
```
Comment (eb8680) [inline effectful/handlers/llm/type_checking.py]: I'm worried that `_type_to_annotation_str` and its subroutine `_collect` are unsound and will silently lead to lots of typechecking failures where the types are rejected as malformed by the typechecker. I think we need to test this function much more systematically.

File: effectful/handlers/llm/type_checking.py [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+            ann_str, imps, type_aliases = _type_to_annotation_str(value)
+            all_imports.extend(imps)
+            all_type_aliases.extend(type_aliases)
+            alias_name = getattr(value, "__name__", name)
+            if alias_name != name:
+                prelude_lines.append(f"{name}: {alias_name}")
+        else:
+            try:
+                inferred = nested_type(value).value
+            except Exception:
```
Comment (eb8680) [inline effectful/handlers/llm/type_checking.py]: Can this type be more specific than `Exception`?

**Review by eb8680 — COMMENTED** [review]
This is nice, but I'm worried about soundness because bugs here will silently cause code generation to fail for no good reason.

It would be ideal if we could define one `singledispatch`-extensible function `quote_type` replacing/subsuming `_type_to_annotation_str` and `_collect` and an equation it satisfies together with `eval`/`exec` (and with the typechecker API), and have it live in `effectful.internals.unification` with tons of parameterized tests covering corner cases, similar to the other generic type-munging functions there.

File: effectful/handlers/llm/type_checking.py [hunk 4]
```diff
@@ -0,0 +1,243 @@
+"""Type-check synthesized code with mypy. Supports Python 3.12 and 3.13."""
+
+import ast
+import collections.abc
+import types
+import typing
+from typing import Any
+
+import mypy.api
```
Comment (eb8680) [inline effectful/handlers/llm/type_checking.py]: I think `mypy` now needs to be added to the `llm` dependency section in `pyproject.toml`

File: effectful/handlers/llm/type_checking.py [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+
+from effectful.internals.unification import nested_type
+
+# PEP 695 (3.12+); use getattr for forward compatibility
+_TypeAliasType = getattr(typing, "TypeAliasType", type(None))
```
[the remark this answers is not in the document]
Comment (kiranandcode) [reply effectful/handlers/llm/type_checking.py]: oh, yep updated.

**Comment by kiranandcode** [comment]
ah but I see your point @eb8680 including sys.modules as well should ensure that all types of values would be present. Though sys.modules is really big.

File: effectful/handlers/llm/evaluation.py [hunk 6]
```diff
@@ -62,10 +92,493 @@ def exec(
...
+        + list(stubs)
+        + list(variables)
+        + list(module.body)
+        + [postlude]
+    )
+    stub_module = ast.Module(body=full_body, type_ignores=[])
+    source = ast.unparse(ast.fix_missing_locations(stub_module))
+    source = (
+        subprocess.run(
+            ["ruff", "check", "--select", "I,F401", "--fix", "--exit-zero", "-"],
```
[the remark this answers is not in the document]
Comment (kiranandcode) [reply effectful/handlers/llm/evaluation.py]: Is this based on the outputs from the failing notebook tests? I believe the imports are pruned, the type checking runs substantially faster. I think the openai and litellm imports are kept because they are needed maybe?

[the remark this answers is not in the document]
Comment (eb8680) [reply effectful/handlers/llm/evaluation.py]: They're not needed though? I mean, maybe top-level `openai` and `litellm` imports are, but there are hundreds of lines of subpackage imports that are unused and that will almost certainly confuse the LLM when it receives an error message from `mypy` that includes them. I think `ruff` is just afraid to delete them because subpackage imports can have side effects?

[the remark this answers is not in the document]
Comment (eb8680) [reply effectful/handlers/llm/evaluation.py]: I would expect all of the imports above to be pruned except for `collections.abc`. For some reason that's not happening.

[the remark this answers is not in the document]
Comment (kiranandcode) [reply effectful/handlers/llm/evaluation.py]:
```python
import completions
import completions.abc

def foo(...):
...
```

This is what would be achieved if we did a while loop on `ruff check --fix`, but would take several iterations.

**Comment by kiranandcode** [comment]
Tests pass!!!!!!!!
